### PR TITLE
load usd prices from data service

### PR DESCRIPTION
**Pull request by domnikov-timofei**

### CODE REVIEW

File: src/ui/components/ui/UsdAmount/index.tsx [hunk 1]
```diff
@@ -8,18 +12,22 @@ interface Props {
...
     return null;
   }
 
-  return !usdPrices[id] || usdPrices[id] === '1' ? null : (
-    <p className={className}>{`≈ $${new BigNumber(usdPrices[id])
-      .mul(tokens)
-      .toFixed(2)}`}</p>
+  return (
+    <p className={className}>
+      {`≈ $${new BigNumber(usdPrices[id]).mul(tokens).toFixed(2)}`}
```
Comment (faergeek) [inline src/ui/components/ui/UsdAmount/index.tsx:30]: Seems like using jsx for formatting would be easier to understand here

File: src/ui/components/ui/UsdAmount/index.tsx [hunk 2]
```diff
@@ -8,18 +12,22 @@ interface Props {
...
   const isMainnet = currentNetwork === 'mainnet';
 
-  if (!usdPrices || !isMainnet) {
+  const usdPrices = useUsdPrices(useMemo(() => [id], [id]));
+
+  if (typeof usdPrices[id] !== 'string') {
+    return <Loader />;
+  }
+
+  if (!isMainnet || !usdPrices[id] || usdPrices[id] === '1') {
```
Comment (faergeek) [inline src/ui/components/ui/UsdAmount/index.tsx:24]: You already check for `usdPrices[id]` above. Checking that it's a string is enough. If it's empty, then it's a backend problem. And we don't need a special case for '1' value, we decided to show usd prices for all assets always.

File: src/controllers/assetInfo.ts [hunk 3]
```diff
@@ -4,7 +4,10 @@ import { NetworkName } from 'networks/types';
 import ObservableStore from 'obs-store';
 import Browser from 'webextension-polyfill';
 
-import { defaultAssetTickers, stablecoinAssetIds } from '../assets/constants';
+import {
+  defaultAssetTickers,
+  // stablecoinAssetIds
```
Comment (faergeek) [inline src/controllers/assetInfo.ts:9]: Commented out import, should be removed

File: src/controllers/assetInfo.ts [hunk 4]
```diff
@@ -392,49 +381,36 @@ export class AssetInfoController {
...
-              ticker.priceAssetID === MARKETDATA_USD_ASSET_ID
-            ) {
-              acc[ticker.amountAssetID] = ticker['24h_close'];
-            }
-
-            return acc;
-          },
-          {}
-        );
+    if (!assetIds.length || network !== NetworkName.Mainnet) {
```
Comment (faergeek) [inline src/controllers/assetInfo.ts:387]: I know that `!assetIds.length` and `assetIds.length === 0` check the same thing, but with the latter the intent is much clearer, don't you agree? :-)

File: src/_core/polling.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
 export function startPolling(
   ms: number,
-  fn: (abortSignal: AbortSignal) => Promise<void>
+  fn: (abortSignal: AbortSignal) => Promise<unknown>
```
Comment (faergeek) [inline src/_core/polling.ts:3]: Is this change really needed? We don't expect function to return anything, because it's a side-effect function, so I think it's better to stick to `void` here

File: src/_core/usdPrices.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+} from 'react';
+import invariant from 'tiny-invariant';
+
+import { usePopupSelector } from '../popup/store/react';
+import Background from '../ui/services/Background';
+import { startPolling } from './polling';
+import { useDebouncedValue } from './useDebouncedValue';
+
+const UsdPricesContext = createContext<
+  null | ((assetIds: string[]) => undefined | (() => void))
```
Comment (faergeek) [inline src/_core/usdPrices.tsx:17]: Values like `null` & `undefined` seems to be much more expected at the end of a union, not at the beginning)

File: src/_core/usdPrices.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+>(null);
+
+export function UsdPricesProvider({ children }: { children: ReactNode }) {
+  const [observedAssetIds, setObservedAssetIds] = useState<string[][]>([]);
+  const observedAssetIdsDebounced = useDebouncedValue(observedAssetIds, 100);
+
+  useEffect(() => {
+    if (!observedAssetIdsDebounced.length) {
```
Comment (faergeek) [inline src/_core/usdPrices.tsx:25]: Same thing, `!<something>.length` vs `<something>length === 0`

File: src/_core/usdPrices.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      return;
+    }
+
+    return startPolling(5000, () =>
+      Background.updateUsdPricesByAssetIds([
+        ...new Set(observedAssetIdsDebounced.flat()),
```
Comment (faergeek) [inline src/_core/usdPrices.tsx:31]: I doubt any other items would ever be added to this array, so may be use `Array.from` instead of spreading onto an empty array?

File: src/_core/usdPrices.tsx [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  const observe = useCallback((assetIds: string[]) => {
+    if (!assetIds.length) {
+      return;
+    }
+
+    setObservedAssetIds(ids => [...ids, assetIds]);
+
+    return () => {
+      setObservedAssetIds(ids => {
+        ids.splice(ids.indexOf(assetIds), 1);
```
Comment (faergeek) [inline src/_core/usdPrices.tsx:45]: Mutating state doesn't seem to be a good idea. May be use `.filter` instead?

File: src/_core/usdPrices.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  invariant(observeUsdPricesByAssetIds);
+
+  const usdPrices = usePopupSelector(state => state.usdPrices);
+
+  useEffect(
+    () => observeUsdPricesByAssetIds(assetIds),
+    [observeUsdPricesByAssetIds, assetIds]
+  );
+
+  return usdPrices;
```
Comment (faergeek) [inline src/_core/usdPrices.tsx:70]: May be filter returned object to only have `assetIds` keys? Otherwise you can write `useUsdPrices([])` and get all the values, except they won't be kept fresh. Which in turn can be a debugging nightmare.

And probably wrap it into a `useMemo` then.

**Review by faergeek — CHANGES_REQUESTED** [review]

**Review by smelnikov — CHANGES_REQUESTED** [review]

**Review by faergeek — APPROVED** [review]

File: src/ui/components/ui/UsdAmount/index.tsx [hunk 11]
```diff
@@ -8,18 +12,22 @@ interface Props {
...
   const isMainnet = currentNetwork === 'mainnet';
 
-  if (!usdPrices || !isMainnet) {
+  const usdPrices = useUsdPrices(useMemo(() => [id], [id]));
+
+  if (typeof usdPrices[id] !== 'string') {
+    return <Loader />;
+  }
+
+  if (!isMainnet || usdPrices[id] === '1') {
```
[the remark this answers is not in the document]
Comment (domnikov-timofei) [reply src/ui/components/ui/UsdAmount/index.tsx]: it was already here, but yes, it seems unnecessary

**Review by faergeek — APPROVED** [review]

**Review by smelnikov — APPROVED** [review]
